File: app/src/main/java/joebruckner/lastpick/UI/MovieFragment.java
package joebruckner.lastpick.ui;


import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.graphics.Palette;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;

import butterknife.Bind;
import butterknife.ButterKnife;
import joebruckner.lastpick.LastPickApp;
import joebruckner.lastpick.R;
import joebruckner.lastpick.data.Movie;
import joebruckner.lastpick.data.MovieViewHolder;
import joebruckner.lastpick.presenters.MoviePresenter;
import joebruckner.lastpick.presenters.MoviePresenterImpl;
import joebruckner.lastpick.widgets.OnAnimationFinishedListener;


public class MovieFragment extends Fragment implements
        MoviePresenter.MovieView, RequestListener<String, Bitmap> {
	@Bind(R.id.content) View content;
	@Bind(R.id.poster) ImageView poster;

	MoviePresenter presenter;
	Coordinator coordinator;
	MovieViewHolder holder;

	private Animation enter;
	private Animation exit;

    private boolean isLoading = true;

	@Override public void onAttach(Context context) {
		super.onAttach(context);
		enter = AnimationUtils.loadAnimation(context, R.anim.slide_up);
		enter.setAnimationListener(new OnAnimationFinishedListener() {
			@Override public void onAnimationEnd(Animation animation) {
				onAnimationFinished(animation);
			}
		});
		exit = AnimationUtils.loadAnimation(context, R.anim.slide_down);
		exit.setAnimationListener(new OnAnimationFinishedListener() {
			@Override public void onAnimationEnd(Animation animation) {
				onAnimationFinished(animation);
			}
		});
		if (context instanceof Coordinator)
			coordinator = (Coordinator) context;

        LastPickApp app = (LastPickApp) getActivity().getApplication();
        presenter = new MoviePresenterImpl(app.getBus());
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		return inflater.inflate(R.layout.fragment_movie, container, false);
	}

	@Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		ButterKnife.bind(this, view);
		holder = new MovieViewHolder(view);
		presenter.attachActor(this);
		presenter.shuffleMovie();
		coordinator.getFab().setOnClickListener(new View.OnClickListener() {
			@Override public void onClick(View v) {
				presenter.shuffleMovie();
			}
		});
	}

	@Override public void showLoading() {
        isLoading = true;
		coordinator.getFab().setClickable(false);
		content.startAnimation(exit);
	}

	@Override public void showContent(Movie movie) {
        isLoading = false;
		Log.d(this.getClass().getSimpleName(), "Showing Movie " + movie.toString());
		Glide.with(this).load(movie.getBackdropPath())
				.asBitmap()
				.listener(this)
				.centerCrop()
				.into(poster);
		Glide.with(this).load(movie.getPosterPath())
				.centerCrop()
				.into(coordinator.getBackdrop());
		holder.setData(movie);
		if (exit.hasStarted()) {
			exit.setAnimationListener(new OnAnimationFinishedListener() {
				@Override public void onAnimationEnd(Animation animation) {
					content.startAnimation(enter);
				}
			});
		} else content.startAnimation(enter);
	}

	@Override public void showError(String errorMessage) {
        isLoading = false;
		Snackbar.make(coordinator.getRootView(), errorMessage, Snackbar.LENGTH_LONG).show();
	}

	protected void onAnimationFinished(Animation animation) {
		if (animation.equals(enter))
			coordinator.getFab().setClickable(true);
	}

	@Override
	public boolean onException(Exception e, String model, Target<Bitmap> target,
	                           boolean isFirstResource) {
		Log.e("Glide", e.toString());
		return false;
	}

	@Override
	public boolean onResourceReady(Bitmap resource, String model, Target<Bitmap> target,
	                               boolean isFromMemoryCache, boolean isFirstResource) {
		Palette.from(resource).generate(new Palette.PaletteAsyncListener() {
			@Override public void onGenerated(Palette palette) {
				generateTheme(palette);
			}
		});
		return false;
	}

	private void generateTheme(Palette palette) {
		Palette.Swatch muted = palette.getMutedSwatch();
		Palette.Swatch vibrant = palette.getVibrantSwatch();
		int primary = Color.parseColor("#455A64");
		int dark = Color.parseColor("#263238");
		int accent = Color.parseColor("#607D8B");
		if (muted != null) {
			Log.d("Palette", "Muted swatch used");
			primary = palette.getMutedColor(Color.BLACK);
			float[] hsl = muted.getHsl();
			hsl[2] *= 0.9;
			dark = Color.HSVToColor(hsl);
			hsl[2] *= 1.8;
			accent = Color.HSVToColor(hsl);
		} else if (vibrant != null) {
			Log.d("Palette", "Vibrant swatch used");
			primary = palette.getVibrantColor(Color.BLACK);
			float[] hsl = vibrant.getHsl();
			hsl[2] *= 0.9;
			dark = Color.HSVToColor(hsl);
			hsl[2] *= 1.8;
			accent = Color.HSVToColor(hsl);
		} else Log.d("Palette", "Default swatch used");
		coordinator.setThemeColors(primary, dark, accent);
		poster.setBackgroundColor(primary);
	}

    @Override
    public boolean isLoading() {
        return isLoading;
    }
}

File: app/src/main/java/joebruckner/lastpick/UI/Home/MovieFragment.kt
package joebruckner.lastpick.ui.home

import android.graphics.Bitmap
import android.os.Bundle
import android.support.v7.graphics.Palette
import android.util.Log
import android.view.View
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.squareup.otto.Bus
import joebruckner.lastpick.LastPickApp
import joebruckner.lastpick.R
import joebruckner.lastpick.data.Movie
import joebruckner.lastpick.events.Action
import joebruckner.lastpick.presenters.MoviePresenter
import joebruckner.lastpick.presenters.MoviePresenterImpl
import joebruckner.lastpick.ui.MovieViewHolder
import joebruckner.lastpick.ui.common.BaseFragment
import joebruckner.lastpick.widgets.ImageBlur
import joebruckner.lastpick.widgets.PaletteMagic
import kotlinx.android.synthetic.fragment_movie.*

class MovieFragment() : BaseFragment(),
        MoviePresenter.MovieView, RequestListener<String, Bitmap> {
    override val layoutId = R.layout.fragment_movie
    override var isLoading = true
    lateinit var holder: MovieViewHolder
    var presenter: MoviePresenter? = null

    lateinit var poster: ImageView
    lateinit var backdrop: ImageView
    lateinit var blur: BitmapTransformation

    override fun showLoading() {
        isLoading = true
        clearMovie()
        parent.disableFab()
        updateViews(View.INVISIBLE, View.VISIBLE, View.INVISIBLE)
    }

    override fun showError(errorMessage: String) {
        isLoading = false
        clearMovie()
        parent.enableFab()
        error.text = errorMessage
        updateViews(View.INVISIBLE, View.INVISIBLE, View.VISIBLE)
    }

    override fun showContent(movie: Movie) {
        isLoading = false
        showMovie(movie)
        parent.enableFab()
        updateViews(View.VISIBLE, View.INVISIBLE, View.INVISIBLE)
    }

    private fun clearMovie() {
        backdrop.setImageResource(android.R.color.transparent)
        poster.setImageResource(android.R.color.transparent)
        parent.title = ""
    }

    private fun showMovie(movie: Movie) {
        holder.movie = movie
        parent.title = movie.title
        setBackdrop(movie.fullBackdropPath())
        setPoster(movie.fullPosterPath())
    }

    private fun updateViews(contentState: Int, loadingState:Int, errorState: Int) {
        if (view == null) return
        content.visibility = contentState
        loading.visibility = loadingState
        error.visibility   = errorState
    }

    fun setBackdrop(imagePath: String) {
        Glide.with(this).load(imagePath)
                .asBitmap()
                .listener(this)
                .transform(blur)
                .into(backdrop)
    }

    fun setPoster(imagePath: String) {
        Glide.with(this).load(imagePath)
                .crossFade()
                .into(poster)
    }

    override fun onException(e: Exception, model: String,
                             target: Target<Bitmap>, isFirstResource: Boolean): Boolean {
        Log.e("Glide", e.toString());
        return false;
    }

    override fun onResourceReady(resource: Bitmap, model: String, target: Target<Bitmap>,
                                 isFromMemoryCache: Boolean, isFirstResource: Boolean): Boolean {
        Palette.from(resource).generate { palette ->
            val magic = PaletteMagic(palette)
            parent.setTheme(magic.primary, magic.dark, magic.accent)
        }
        return false
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        parent.setToolbarStubLayout(R.layout.backdrop_movie)
        holder = MovieViewHolder(view)

        backdrop = parent.root.findViewById(R.id.backdrop) as ImageView
        poster = parent.root.findViewById(R.id.poster) as ImageView

        blur = ImageBlur(context, 20f)
        parent.appBar.addOnOffsetChangedListener { layout, i ->
            poster.elevation = if(layout.y < -40) 0f else 4f
            poster.alpha = (100 + layout.y) / 100
        }

        menuId = R.menu.menu_history
        val bus = parent.application.getSystemService(LastPickApp.BUS) as Bus
        presenter = MoviePresenterImpl(bus)
        presenter?.attachActor(this)
    }

    override fun onResume() {
        super.onStart()
        presenter?.attachActor(this)
    }

    override fun onPause() {
        presenter?.detachActor()
        super.onPause()
    }

    override fun handleAction(action: Action) {
        when (action.name) {
            Action.UPDATE -> presenter?.updateMovie()
        }
    }
}